dpcpp/components/cooperative_groups.dp.hpp

```diff
@@ -0,0 +1,513 @@
+ *       to existing cooperative groups, or create new groups if the existing
+ *       groups do not cover your use-case. For an example, see the
+ *       enable_extended_shuffle mixin, which adds extended shuffles support
+ *       to built-in DPCPP cooperative groups.
+ */
+namespace group {
+
+
+// metafunctions
+namespace detail {
+
+
+template <typename T>
+struct is_group_impl : std::false_type {};
+
+
+template <typename T>
+struct is_synchronizable_group_impl : std::false_type {};
+
+
+template <typename T>
+struct is_communicator_group_impl : std::true_type {};
+
+
+}  // namespace detail
+
+
+/**
+ * Check if T is a Group.
+ */
+template <typename T>
+using is_group = detail::is_group_impl<std::decay_t<T>>;
+
+
+/**
+ * Check if T is a SynchronizableGroup.
+ */
+template <typename T>
+using is_synchronizable_group =
+    detail::is_synchronizable_group_impl<std::decay_t<T>>;
+
+
+/**
+ * Check if T is a CommunicatorGroup.
+ */
+template <typename T>
+using is_communicator_group =
+    detail::is_communicator_group_impl<std::decay_t<T>>;
+
+
+// types
+namespace detail {
+
+
+/**
+ * This is a limited implementation of the DPCPP thread_block_tile.
+ */
+template <unsigned Size>
+class thread_block_tile : public cl::sycl::ONEAPI::sub_group {
+    using sub_group = cl::sycl::ONEAPI::sub_group;
+    using id_type = sub_group::id_type;
+    using mask_type = config::lane_mask_type;
+
+public:
+    // note: intel calls nd_item.get_sub_group(), but it still call
+    // intel::sub_group() to create the sub_group.
+    template <typename Group>
+    explicit thread_block_tile(const Group &parent_group)
+        : data_{Size, 0}, sub_group()
+    {
+#ifndef NDEBUG
+        assert(this->get_local_range().get(0) == Size);
+#endif
+        data_.rank = this->get_local_id();
+    }
+
+
+    __dpct_inline__ unsigned thread_rank() const noexcept { return data_.rank; }
+
+    __dpct_inline__ unsigned size() const noexcept { return Size; }
+
+    __dpct_inline__ void sync() const noexcept {}
+
+#define GKO_BIND_SHFL(ShflOpName, ShflOp)                                      \
+    template <typename ValueType, typename SelectorType>                       \
+    __dpct_inline__ ValueType ShflOpName(ValueType var, SelectorType selector) \
+        const noexcept                                                         \
+    {                                                                          \
+        return this->ShflOp(var, selector);                                    \
+    }                                                                          \
+    static_assert(true,                                                        \
+                  "This assert is used to counter the false positive extra "   \
+                  "semi-colon warnings")
+
+    GKO_BIND_SHFL(shfl, shuffle);
+    GKO_BIND_SHFL(shfl_xor, shuffle_xor);
+
+    // the shfl_up of out-of-range value gives undetermined behavior, we
```

Comment:
What is the wording here? Does it give *undefined behavior*, or return *undefined* or *implementation-defined* values? That might be interesting for algorithms where we cause this overflow on purpose because we discard the values anyways.

Reply:
it is undefined behavior. we do not use it in algorithm actually, but we use it to pass cooperative group test